python/groupby.cpp

```diff
@@ -6,59 +6,64 @@
 #include "scipp/dataset/groupby.h"
 #include "scipp/dataset/dataset.h"
 
+#include "bind_free_function.h"
 #include "pybind11.h"
 
 using namespace scipp;
 using namespace scipp::dataset;
+using namespace scipp::python;
 
 namespace py = pybind11;
 
 template <class T> void bind_groupby(py::module &m, const std::string &name) {
-  m.def("groupby",
-        py::overload_cast<const typename T::const_view_type &, const Dim>(
-            &groupby),
-        py::arg("data"), py::arg("group"),
-        py::call_guard<py::gil_scoped_release>(),
-        R"(
-        Group dataset or data array based on values of specified labels.
-
-        :param data: Input dataset or data array
-        :param group: Name of labels to use for grouping
-        :type data: DataArray or Dataset
-        :type group: str
-        :return: GroupBy helper object.
-        :rtype: GroupByDataArray or GroupByDataset)");
-  m.def("groupby",
-        py::overload_cast<const typename T::const_view_type &, const Dim,
-                          const VariableConstView &>(&groupby),
-        py::arg("data"), py::arg("group"), py::arg("bins"),
-        py::call_guard<py::gil_scoped_release>(),
-        R"(
-        Group dataset or data array based on values of specified labels.
-
-        :param data: Input dataset or data array
-        :param group: Name of labels to use for grouping
-        :param bins: Bins for grouping label values
-        :type data: DataArray or Dataset
-        :type group: str
-        :type bins: VariableConstView
-        :return: GroupBy helper object.
-        :rtype: GroupByDataArray or GroupByDataset)");
+
+  using VAConstView = const typename Variable::const_view_type &;
+  using ConstView = const typename T::const_view_type &;
+  Docstring docs;
+
+  // Flatten
+  docs = {// Description
+          "Group dataset or data array based on values of specified labels.",
+          // Raises
+          "",
+          // See also
+          "",
+          // Returns
+          "GroupBy helper object.",
+          // Return type
+          // "GroupByDataArray or GroupByDataset.",
+          // Input parameters
+          {{"data", "Input Dataset or DataArray."},
+           {"group", "String containing name of labels to use for grouping."}}};
+  // Overload on function that accepts docstring is ambiguous, use long version
+  // of constructor
+  bind_free_function<GroupBy<T>, ConstView, const Dim>(
+      groupby, "groupby", m, docs.param(0), docs.param(1), docs.description(),
+      docs.raises(), docs.seealso(), docs.returns());
```

Comment:
I really appreciate the effort and the intentions, but to be brutally honest I feel this is making things worse:

` Constructing `Docstring` is counterintuitive and error prone, an issue highlighted by the need to put comments in the constructor call.
- Constructing `Docstring` is very long.
- `docs.param(0), docs.param(1)` is less readable than the native `py::arg("dim")` where the reader sees what is being defined.
- This solution forces developers to **learn a new syntax**. Knowledge of `pybind11` (or `boost::python`, which is similar) does not help, thus raising the barrier of entry and increasing the cognitive burden for existing developers.
- If we want to change `Docstring`, such as by adding or removing a field later we need to go back and update *all* bindings.
- More than 100 lines of code to define `bind_free_function`.
- Things like `docs.insert_param(2, {"bins", "Bins for grouping label values."});` to be able to reuse a docstring look error prone, but more importantly points at the problem I mentioned in the past: the issue with the "overloaded" function documentation (see https://scipp.github.io/python-reference/api.html#free-functions which is not very useful and us even worse subpages listing all the overloads with lots of repetition). Python does not really have overloaded functions, so we would much rather have a single binding/docstring. *Implementing the solution proposed here will tie us harder to this current suboptimal way of multiple docstrings.* Maybe we will want to switch to only a single binding that supports all "overloads" internally in the future.

Reply:
I see your points, and of course I was expecting this to turn into a discussion. Just wanted to get it in a state where I was relatively close to having it working almost everywhere, to be sure I thought of most cases.

> * `Constructing`Docstring` is counterintuitive and error prone, an issue highlighted by the need to put comments in the constructor call.
I agree on this, and realised that probably a better and clearer solution should be found for the `Docstring` constructor.
> * Constructing `Docstring` is very long.
I'm not so sure how much longer it is, compared to having to write long docstrings, multiple times, which are almost the same. It is also useful in keeping the content of the docstrings standardised across the different functions, as some only had a description and a return type, while others had a description for each parameter, `:raises:`. `:seealso:`.
> * `docs.param(0), docs.param(1)` is less readable than the native `py::arg("dim")` where the reader sees what is being defined.
Yes, i'm not so happy with having these hard coded numbers. If it's a matter of being explicit, maybe a better name might help? (`docs.pyargs`?). Also, `docs` is not a great name either, as you think of `docs` the documentation folder.
> * This solution forces developers to **learn a new syntax**. Knowledge of `pybind11` (or `boost::python`, which is similar) does not help, thus raising the barrier of entry and increasing the cognitive burden for existing developers.
I'm not sure how difficult this is for current developers. It was actually to make their life easier, by not having to worry about some `pybind11` specialities, such as when to use a `lambda` and when not to (in this case I assumed that the lambda would work in all cases), or to remember that you need `py::call_guard<py::gil_scoped_release>()`
> * If we want to change `Docstring`, such as by adding or removing a field later we need to go back and update _all_ bindings.
Yes, this is a big problem.
> * More than 100 lines of code to define `bind_free_function`.
This is not a big issue in my opinion, unless you need to modify it all the time.
> * Things like `docs.insert_param(2, {"bins", "Bins for grouping label values."});` to be able to reuse a docstring look error prone, but more importantly points at the problem I mentioned in the past: the issue with the "overloaded" function documentation (see https://scipp.github.io/python-reference/api.html#free-functions which is not very useful and us even worse subpages listing all the overloads with lots of repetition). Python does not really have overloaded functions, so we would much rather have a single binding/docstring. _Implementing the solution proposed here will tie us harder to this current suboptimal way of multiple docstrings._ Maybe we will want to switch to only a single binding that supports all "overloads" internally in the future (maybe using `py::object`, or `py::args` and py::kwargs`).
Yes I haden't tried to tackle the `"overloaded function"` problem yet, but I didn't think that doing things this way would make it more difficult to fix compared to what we had previously.
Again, I didn't like the `insert_param` with hard-coded numbers, and was hoping someone might have a better idea.

Reply:
> I see your points, and of course I was expecting this to turn into a discussion. Just wanted to get it in a state where I was relatively close to having it working almost everywhere, to be sure I thought of most cases.
> 
> > * `Constructing`Docstring` is counterintuitive and error prone, an issue highlighted by the need to put comments in the constructor call.
> 
> I agree on this, and realised that probably a better and clearer solution should be found for the `Docstring` constructor.

I don't think you can really get around this problem without keyword arguments.

> > * Constructing `Docstring` is very long.
> 
> I'm not so sure how much longer it is, compared to having to write long docstrings, multiple times, which are almost the same.

We can reuse the same string or use a template, e.g., as done here: https://github.com/scipp/scipp/blob/d218f1c1a57f374c11157eb7b48a57657e02901d/python/neutron.cpp#L17-L22.

> It is also useful in keeping the content of the docstrings standardised across the different functions, as some only had a description and a return type, while others also had a description for each parameter, `:raises:`. `:seealso:`, etc.

I agree, but this could be solved by simply having a helper function to write a "standard" docstring.

> > * `docs.param(0), docs.param(1)` is less readable than the native `py::arg("dim")` where the reader sees what is being defined.
> 
> Yes, i'm not so happy with having these hard coded numbers. If it's a matter of being explicit, maybe a better name might help? (`docs.pyargs`?). Also, `docs` is not a great name either, as you think of `docs` the documentation folder.

I would suggest to not do this at all, but stick with `py::arg`. I really do think have to memory two syntaxes to do the same things is too confusing and too much of a burden. The API you are introducing is not fully replacing pybind11, e.g., you still need to native syntax for everything that is not a free function, and two APIs to do the same thing are not maintainable, imho.


> > * This solution forces developers to **learn a new syntax**. Knowledge of `pybind11` (or `boost::python`, which is similar) does not help, thus raising the barrier of entry and increasing the cognitive burden for existing developers.
> 
> I'm not sure how difficult this is for current developers. It was actually to make their life easier, by not having to worry about some `pybind11` specialities, such as when to use a `lambda` and when not to (in this case I assumed that the lambda would work in all cases), or to remember that you need `py::call_guard<py::gil_scoped_release>()`

Well, I don't think you can solve the speciality. The solution here is basically "use `py::overload_cast` (that is essentially what the explicit template arguments to `bind_free_function` are, compare the syntax). If the variety is a problem we can standardize on always using a lambda. The GIL release is a fair point, but it is also risky: See a couple of special cases where the GIL *cannot* be release right away (such as `realign`). So either `bind_free_function` cannot be used, or you need to have special arguments to disable it. In any case, it will be non-obvious to the reader of the bindings that the GIL is released, without looking into the implementation (or documentation).


> > * If we want to change `Docstring`, such as by adding or removing a field later we need to go back and update _all_ bindings.
> 
> Yes, this is a big problem. Could maybe be solved if we have a `kwargs` syntax like Owen suggests above?
> 
> > * More than 100 lines of code to define `bind_free_function`.
> 
> This is not a big issue in my opinion, unless you need to modify it all the time. But as I mentioned, it would be nice if a syntax could be found where it did not depend on the number of the function arguments, thus also removing the need to add a function for 5 or 6 arguments if new functions require that.

I agree that this is not an issue *now*. But looking, e.g., at Mantid you can see that it just kept on growing and growing, and 10 years later it suddenly *is* a problem that cannot be solved since it is so tied into itself.

> > * Things like `docs.insert_param(2, {"bins", "Bins for grouping label values."});` to be able to reuse a docstring look error prone, but more importantly points at the problem I mentioned in the past: the issue with the "overloaded" function documentation (see https://scipp.github.io/python-reference/api.html#free-functions which is not very useful and us even worse subpages listing all the overloads with lots of repetition). Python does not really have overloaded functions, so we would much rather have a single binding/docstring. _Implementing the solution proposed here will tie us harder to this current suboptimal way of multiple docstrings._ Maybe we will want to switch to only a single binding that supports all "overloads" internally in the future (maybe using `py::object`, or `py::args` and py::kwargs`).
> 
> Yes I haden't tried to tackle the `"overloaded function"` problem yet, but I didn't think that doing things this way would make it more difficult to fix compared to what we had previously.

Not "difficult", but time consuming: (1) You end up rewriting things twice, which wastes effort, (2) The new syntax is more entangled and harder to refactor.


Can't we solve 80% of what you intent to solve by using the linked approach, or by adding a couple of helper functions to write docstrings, without all the significant drawbacks discussed above?

Reply:
So the reason why I tried to go beyond the linked templated approach was to try and see if we could remove/minimise the need for adding things in more than one place when adding a new function binding. Having to write a template first at the top of the file, and then a call further down.

I agree it's not a terrible nuisance, but thought it might be useful and limit errors/oversights if we could have it all in the same code block.

For the Docstring constructor, what do you think of the solution I suggested above in a reply to Owen's comment, something like `Docstring().description("Element-wise absolute value.").raises("If input is integer")...`?